### ISSUE TITLE
Parsing of layer name from CSW

### ISSUE THREAD

**Issue by tdipisa** [geosolutions-it/MapStore2-C027#72]
Update of the CSW catalog parsing in MS to retrieve the layer name also from the URL of the resource in cases where the name property is missing in the metadata. This is due to the last RNDT / INSPIRE specifications.

**Comment by tdipisa** [comment]
@randomorder it seems not possible to perform HTTP requests from sr-vm377-sitms to sr-vm378-sitgfn: the wget got stuck:

![image](https://user-images.githubusercontent.com/1280027/136993847-0f21fd78-d1f7-4a26-87e7-174fdb2bdf09.png)

What I'm trying to do is to reach the CSW [here](http://sr-vm378-sitgfn.comune.intranet:9080/geonetwork/srv/eng/csw) from the catalog tool of MS and the request is passing through the MS proxy

![image](https://user-images.githubusercontent.com/1280027/136994268-3a37d97a-57d7-4a46-85b2-962d68ee7f67.png)

**Comment by randomorder** [comment]
How are the env vars on the command line and java opts for mapstore set? I am interested in proxy settings

**Comment by lpasquali** [comment]
these are the resulting expanded parameters inb the linux process, in my opinion there is a good use of proxy values for firenze network, so that proxy is avoided for internal services:
```
/usr/lib/jvm/java-8-openjdk-amd64/bin/java -Djava.util.logging.config.file=/var/lib/tomcat9/conf/logging.properties -Djava.util.logging.manager=org.apache.juli.ClassLoaderLogManager -Djava.awt.headless=true -Dhttp.proxyHost=proxyhttp.comune.intranet -Dhttp.proxyPort=8080 -Dhttps.proxyHost=proxyhttps.comune.intranet -Dhttps.proxyPort=8080 -Dhttp.nonProxyHosts=*.comune.intranet|192.168.*|localhost -Dgeostore-ovr=file:///var/lib/mapstore2/conf/geostore-datasource-ovr.properties -Xms512M -Xmx1024M -server -XX:+UseParallelGC -Djava.awt.headless=true -Djdk.tls.ephemeralDHKeySize=2048 -Djava.protocol.handler.pkgs=org.apache.catalina.webresources -Dorg.apache.catalina.security.SecurityListener.UMASK=0027 -Dignore.endorsed.dirs= -classpath /usr/share/tomcat9/bin/bootstrap.jar:/usr/share/tomcat9/bin/tomcat-juli.jar -Dcatalina.base=/var/lib/tomcat9 -Dcatalina.home=/usr/share/tomcat9 -Djava.io.tmpdir=/tmp org.apache.catalina.startup.Bootstrap start
```

**Comment by lpasquali** [comment]
@randomorder nothing on MS side logs

**Comment by randomorder** [comment]
at the very least the request should be logged with an HTTP status

**Comment by lpasquali** [comment]
@randomorder @tdipisa even if I open at most http-proxy configuration (allow any request, mimetype, host or ip)  I keep getting the same results and no traffic is getting to geonetwork.
I then removed firenze sysadmins' proxy part from JAVA_OPTS, this got geonetwork reachable and correctly configurable in mapstore catalogs
But at this point I see that even if the csw url is added correctly, then I can't search anything but some of the layers in geonetwork catalog by their full name (like "Aree giochi") but result looks like duplicated and lacking of metadata:
![Screenshot from 2021-10-19 12-24-21](https://user-images.githubusercontent.com/1780406/137891766-18a40de1-6913-4232-ad19-cd1ce2818b42.png)

every search on geonetwork results in this warning message:
2021-10-19 12:24:20,207 WARN  [jeeves] - Run in transaction did not complete cleanly, transaction not committed or rolledback

**Comment by tdipisa** [comment]
>- Proxy settings are not working in MapStore and we don't know why but removing them MS is able to reach GeoNetwork.
As long as MapStore is working fine with all the services (including external ones) that is fine. We can carry on an investigation in parallel

@lpasquali @randomorder 

Now MS is fetching results from the new GN therefore, as per the development is concerned, we are good but I think that the proxy issue should be anyway investigated with  Comune FI admins. @randomorder if you prefer I can ask to Carnasciali or someone else of his team to participate to the call tomorrow, let me know.

> but I think that adding an external internet catalog may not be possible if there is not such proxy setting in JAVA_OPTS

Indeed, if I try to connect our GN the result is the follwowing

![image](https://user-images.githubusercontent.com/1280027/138090919-6ac1ffb1-a1a9-4700-bb6d-5c4cd1491b99.png)

I've opened a dedicated issue for this investigation (#76) since this issue has a different purpose.

>- Duplicate layers and missing metadata is another issue

They are not duplicates, it is simply because GN is not returning the metadata in the expected format (ISO instead of DC). I've opened a dedicated issue for this too @etj (#75)

**Comment by lpasquali** [comment]
after the call of last week, we are waiting for Carnasciali to come up with a fix for the proxy

**Comment by tdipisa** [comment]
@etj as we have seen today GN has some issue that are preventing us to define WMS online resources where needed:

![image](https://user-images.githubusercontent.com/1280027/140549733-0ae2680d-fec2-4e85-a4ea-eb1a307df106.png)

We have also to understand why WMS online resources of existing RNDT metadata are not in the right tab (Dataset e serie).

Finally I need to know how to update online resources referring a layer (WMS or WMTS) of existing metadata to be able to parse the layer name MS side as expected by the new RNDT specs.

**Comment by tdipisa** [comment]
Thanks to @etj, I can now properly review and update the online resource for one of the existing metadata (according to new RNDT standard version) to be used as a sample for the development. The metadata is [this](http://sr-vm378-sitgfn.comune.intranet:9080/geonetwork/srv/ita/catalog.search#/metadata/c_d612:d0ffa579-e002-4dc3-94b4-d69207cae114)

Below is the WMS online resource configured with the URL expected by RNDT v2:

![image](https://user-images.githubusercontent.com/1280027/141140306-a31587cd-e69b-4809-9d0b-bcdd8adb6792.png)

DC output MS side reports the corresponding dc:URI element with the newly configured URL but the _protocol_ attribute is not reported as expected in the XML:

![image](https://user-images.githubusercontent.com/1280027/141139929-e161ff2d-3459-4030-a144-924d3e738e55.png)

Below is the DC output of the same metadata from the old GN version currently in production

![image](https://user-images.githubusercontent.com/1280027/141140817-e7cc4e20-d430-476d-9a80-52f0d93167e0.png)

**Comment by etj** [comment]
Ref https://github.com/metadata101/iso19139.rndt/issues/20

Fixing so that the `protocol` attribute will return values from the INSPIRE codelist:

e.g.
```xml
    <dc:URI protocol="http://www.opengis.net/def/serviceType/ogc/wms" description="access point">http://server1/wms?service=wms&amp;request=getCapabilities</dc:URI>
    <dc:URI protocol="http://www.opengis.net/def/serviceType/ogc/wms" description="access point">http://server1/wms?service=wms&amp;request=getMap&amp;layer=pippo</dc:URI>
    <dc:URI protocol="http://www.opengis.net/def/serviceType/ogc/wfs" description="access point">http://server1/wms?service=WFS&amp;request=getCapabilities</dc:URI>
    <dc:URI protocol="http://www.opengis.net/def/serviceType/ogc/wfs" description="access point">http://server1/ows?service=WFS&amp;request=getFeature&amp;typeNames=pippo</dc:URI>
  </csw:Record>
```

**Comment by etj** [comment]
@tdipisa fixes and improvements committed and deployed on C027 server.

**Comment by tdipisa** [comment]
Thank you so much @etj I will check it in MS asap.

**Comment by tdipisa** [comment]
Ok so. I think we can proceed with the development MS side by improving the Catalog tool to be able  parse the new DC output.

The things to review/fix here are at least:
1. Parse the new DC output so that layers (WMS/WMTS online resources) can be added to the MS map. Metadata samples are:
   - With WMS online resource: http://sr-vm378-sitgfn.comune.intranet:9080/geonetwork/srv/ita/catalog.search#/metadata/c_d612:d0ffa579-e002-4dc3-94b4-d69207cae114
   - With WMTS online resource:
2. Fix/review the parsing of new DC output to show metadata info as configured through the [metadata template](https://mapstore.readthedocs.io/en/latest/user-guide/catalog/#metadata-templates). Some of them are no longer well parsed, see screenshot below:

![2021_11_15_16_01_56_](https://user-images.githubusercontent.com/1280027/141804359-e9bc7278-5864-4789-8120-f24130eadea6.png)

@etj the element **dc:contributor** is no longer exposed by the CSW output in DC format

**Comment by etj** [comment]
- CSW record missing resp parties: https://github.com/metadata101/iso19139.rndt/issues/21
- CSW record missing temporal info: https://github.com/metadata101/iso19139.rndt/issues/22

**Comment by tdipisa** [comment]
Thank you @etj. Ok so, @offtherailz I think we can now start involving one of our resources to proceed with the development MS side by improving the Catalog tool so that it will be possible to parse the new DC output.

The things to review/fix here are at least:
1. Parse the new DC output so that layers (WMS or WMTS online resources) can be added to the MS map. Metadata samples are:
   - http://sr-vm378-sitgfn.comune.intranet:9080/geonetwork/srv/ita/catalog.search#/metadata/c_d612:d0ffa579-e002-4dc3-94b4-d69207cae114 (for WMS online resources)
   - TBD (for WMTS resources)
2. Fix/review the parsing of the new DC output to ensure the metadata info are showed as configured through the [metadata template](https://mapstore.readthedocs.io/en/latest/user-guide/catalog/#metadata-templates). Some of them are no longer well parsed, see screenshot below:

![image](https://user-images.githubusercontent.com/1280027/141829955-955121f2-60ad-482e-85d7-a67911a2b7dd.png)

Sample of map used for the above screenshot [here](http://sr-vm377-sitms.comune.intranet:8080/mapstore2/#/viewer/openlayers/3072)

**Comment by ale-cristofori** [comment]
Analysing the response we get back from the CSW service there are a couple of issues which would need clarification before proceeding with the development.

Taking the following XML response as example:

`<?xml version="1.0" encoding="UTF-8"?>
<csw:GetRecordsResponse xmlns:csw="http://www.opengis.net/cat/csw/2.0.2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/cat/csw/2.0.2 http://schemas.opengis.net/csw/2.0.2/CSW-discovery.xsd">
  <csw:SearchStatus timestamp="2021-11-16T17:35:09" />
  <csw:SearchResults numberOfRecordsMatched="6" numberOfRecordsReturned="2" elementSet="full" nextRecord="0">
    <csw:Record xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:dct="http://purl.org/dc/terms/" xmlns:ows="http://www.opengis.net/ows" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:gmx="http://www.isotc211.org/2005/gmx" xmlns:gml="http://www.opengis.net/gml/3.2">
      <dc:identifier>c_d612:db2ccd8e-f2f2-47ba-b3fc-ee9f0b45746f</dc:identifier>
      <dc:date>2021-11-16</dc:date>
      <dc:title>Subaree verdi</dc:title>
      <dct:abstract>Elementi poligonali che rappresentano le subaree in cui sono suddivise le aree verdi. Le subaree sono caratterizzate e distinte da un particolare uso del suolo</dct:abstract>
      <dc:description>Elementi poligonali che rappresentano le subaree in cui sono suddivise le aree verdi. Le subaree sono caratterizzate e distinte da un particolare uso del suolo</dc:description>
      <dc:type>dataset</dc:type>
      <dc:subject>aree verdi</dc:subject>
      <dc:subject>verde pubblico</dc:subject>
      <dc:subject>parchi e giardini</dc:subject>
      <dc:subject>ambiente</dc:subject>
      <dc:subject>Utilizzo del territorio</dc:subject>
      <dc:subject>environment</dc:subject>
      <dc:format>WMS</dc:format>
      <dc:contributor>Direzione Ambiente</dc:contributor>
      <dc:rights>otherRestrictions</dc:rights>
      <dc:rights>otherRestrictions</dc:rights>
      <dc:language>ita</dc:language>
      <dc:source>Il SIT del comune di Firenze è attivo dagli anni 90 e promuove la georeferenziazione dei dati e  la costruzione di archivi geografici. Il SIT attualmente fornisce l'infrastruttura tecnologica e le applicazioni. L'aggiornamento del dato è a cura delle direzioni titolari del dato.</dc:source>
      <ows:BoundingBox crs="urn:ogc:def:crs:http://www.epsg-registry.org::3003">
        <ows:LowerCorner>11.348 43.718</ows:LowerCorner>
        <ows:UpperCorner>11.145 43.84</ows:UpperCorner>
      </ows:BoundingBox>
      <dct:temporal>start=2021-10-04; end=2021-10-04</dct:temporal>
      <dc:URI protocol="https://registry.geodati.gov.it/metadata-codelist/ProtocolValue/www-download" profile="http://inspire.ec.europa.eu/metadata-codelist/SpatialDataServiceType/download" description="access point">http://datigis.comune.fi.it/shp/subaree_verdi.zip</dc:URI>
      <dc:URI protocol="https://registry.geodati.gov.it/metadata-codelist/ProtocolValue/www-download" profile="http://inspire.ec.europa.eu/metadata-codelist/SpatialDataServiceType/view" description="access point">http://datigis.comune.intranet/geoserver/wms?tiled=true&amp;version=1.1.1</dc:URI>
      <dc:URI protocol="http://www.w3.org/TR/xlink/" profile="http://inspire.ec.europa.eu/metadata-codelist/SpatialDataServiceType/other" description="access point">http://datigis.comune.fi.it/kml/subaree_verdi.kmz</dc:URI>
      <dc:URI protocol="image/png" name="thumbnail">http://sr-vm378-sitgfn:9080/geonetwork/srv/api/records/c_d612:db2ccd8e-f2f2-47ba-b3fc-ee9f0b45746f/attachments/subaree_s.png</dc:URI>
      <dc:URI protocol="image/png" name="large_thumbnail">http://sr-vm378-sitgfn:9080/geonetwork/srv/api/records/c_d612:db2ccd8e-f2f2-47ba-b3fc-ee9f0b45746f/attachments/subaree.png</dc:URI>
    </csw:Record>
    <csw:Record xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:dct="http://purl.org/dc/terms/" xmlns:ows="http://www.opengis.net/ows" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:gmx="http://www.isotc211.org/2005/gmx" xmlns:gml="http://www.opengis.net/gml/3.2">
      <dc:identifier>c_d612:d0ffa579-e002-4dc3-94b4-d69207cae114</dc:identifier>
      <dc:date>2021-11-12</dc:date>
      <dc:title>Aree verdi</dc:title>
      <dct:abstract>Elementi areali che individuano sul territorio comunale fiorentino le principali aree verdi</dct:abstract>
      <dc:description>Elementi areali che individuano sul territorio comunale fiorentino le principali aree verdi</dc:description>
      <dc:type>dataset</dc:type>
      <dc:subject>aree verdi</dc:subject>
      <dc:subject>verde pubblico</dc:subject>
      <dc:subject>parchi e giardini</dc:subject>
      <dc:subject>ambiente</dc:subject>
      <dc:subject>Copertura del suolo</dc:subject>
      <dc:subject>environment</dc:subject>
      <dc:format>WMS</dc:format>
      <dc:contributor>Direzione Ambiente</dc:contributor>
      <dc:rights>otherRestrictions</dc:rights>
      <dc:rights>otherRestrictions</dc:rights>
      <dc:language>ita</dc:language>
      <dc:source>Il SIT del comune di Firenze è attivo dagli anni 90 e promuove la georeferenziazione dei dati e  la costruzione di archivi geografici. Il SIT attualmente fornisce l'infrastruttura tecnologica e le applicazioni. L'aggiornamento del dato è a cura delle direzioni titolari del dato.</dc:source>
      <ows:BoundingBox crs="urn:ogc:def:crs:http://www.epsg-registry.org::3003">
        <ows:LowerCorner>11.348 43.718</ows:LowerCorner>
        <ows:UpperCorner>11.145 43.84</ows:UpperCorner>
      </ows:BoundingBox>
      <dct:temporal>start=2021-10-04; end=2021-10-04</dct:temporal>
      <dc:URI protocol="https://registry.geodati.gov.it/metadata-codelist/ProtocolValue/www-download" profile="http://inspire.ec.europa.eu/metadata-codelist/SpatialDataServiceType/download" description="access point">http://datigis.comune.fi.it/shp/aree_verdi.zip</dc:URI>
      <dc:URI protocol="http://www.opengis.net/def/serviceType/ogc/wms" profile="http://inspire.ec.europa.eu/metadata-codelist/SpatialDataServiceType/view" description="access point">http://datigisnew.comune.intranet/geoserver/wms?SERVICE=WMS&amp;VERSION=1.3.0&amp;REQUEST=GetMap&amp;FORMAT=image%2Fpng&amp;TRANSPARENT=true&amp;LAYERS=sivep%3Aaree_verdi&amp;STYLES=&amp;SRS=EPSG%3A3857&amp;CRS=EPSG%3A3857&amp;TILED=true&amp;WIDTH=256&amp;HEIGHT=256&amp;BBOX=1245006.3167089522%2C5430086.489378922%2C1247452.3016140778%2C5432532.474284047</dc:URI>
      <dc:URI protocol="http://www.w3.org/TR/xlink/" profile="http://inspire.ec.europa.eu/metadata-codelist/SpatialDataServiceType/view" description="access point">http://datigis.comune.fi.it/kml/aree_verdi.kmz</dc:URI>
      <dc:URI protocol="image/png" name="thumbnail">http://sr-vm378-sitgfn:9080/geonetwork/srv/api/records/c_d612:d0ffa579-e002-4dc3-94b4-d69207cae114/attachments/areeverdi_s.png</dc:URI>
      <dc:URI protocol="image/png" name="large_thumbnail">http://sr-vm378-sitgfn:9080/geonetwork/srv/api/records/c_d612:d0ffa579-e002-4dc3-94b4-d69207cae114/attachments/areeverdi.png</dc:URI>
    </csw:Record>
  </csw:SearchResults>
</csw:GetRecordsResponse>`

**Comment by etj** [comment]
@ale-cristofori 

> I was wondering if the lack of such information is an error and needs fixing or it is just that piece of information isn't available (i.e. there is not a WMS available for those layers).

If we don't have such information, we can't know if it's an error in the metadata or that information is not there at all.
For instance, the metadata could be documenting a dataset with only downloadable ZIP data, or with a link to an external site which provides the data. In this case, it's correct there are no links to OGC services at all.

To be clear, RNDT only recommends the documentation of the getCapabilties URL; in this case, we do not know which layer should be shown, since the `name` attribute is not used in RNDT specs. 
We are asking our customer using RNDT to also add an onlineResource with protocol=wms and profile=view containing a WMS request that we can parse.
[Here](https://docs.google.com/document/d/1iYwASs_MbuLAmhsfEvuuh1yp0Q9NUZD3h0vM8Cdy8WI/edit#heading=h.5sgb7nvz8a7k) some suggestions given to our customers for editing onlineResources (Italian only).

Another issue in the RNDT specs, is that the `onlineResource` info does not have any human-readable information about the single resource, since the "name" element is not used and the "description" element should be assigned the fixed "access point" value.

**Comment by tdipisa** [comment]
@ale-cristofori thank you for the update. Can you please produce an updated WAR file to be deployed in the client environment? There is no a CI/CD  in place at the moment for this. As soon as you provide the WAR one devop can take care of the deploy @randomorder or @lpasquali I think.

**Comment by ale-cristofori** [comment]
> @ale-cristofori thank you for the improvement. Can you please produce an updated WAR file to be deployed in the client environment?

Sure @tdipisa, following https://mapstore.readthedocs.io/en/v2021.01.00/developer-guide/building-and-deploying/ 
which version identifier shall I use for the deploy?

**Comment by tdipisa** [comment]
@ale-cristofori I think you can simply follow [this](https://github.com/geosolutions-it/MapStore2-C027#to-create-a-deployable-war) as I explained in private chat.

**Comment by ale-cristofori** [comment]
@randomorder @lpasquali -  the WAR file is ready to be deployed. Where shall I share it?

**Comment by randomorder** [comment]
Thanks,
I'll contact you in private @ale-cristofori

**Comment by randomorder** [comment]
new mapstore war installed @tdipisa

**Comment by tdipisa** [comment]
ok, thank you @randomorder.
